### PR TITLE
[Ready] Fixes active response completion of the promise. 

### PR DESCRIPTION
**Pull request by rabbah**
While the active ack promise's future was bounded by timeout, it in fact would not fail as expected by the controller since the promise was never completed. Addressed by completing the promise on timeout (i.e., when active ack expiration has occurred).

IPG 183.

### CODE REVIEW

File: common/scala/src/main/scala/whisk/core/entity/Exec.scala [hunk 1]
```diff
@@ -125,7 +125,7 @@ protected[core] object Exec
             val obj = v.asJsObject
 
             val kind = obj.getFields("kind") match {
-                case Seq(JsString(k)) => k
+                case Seq(JsString(k)) => k.trim.toLowerCase
```
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/Exec.scala]: make case insensitive for greater tolerance.

File: common/scala/src/main/scala/whisk/utils/ExecutionContextFactory.scala [hunk 2]
```diff
@@ -17,22 +17,36 @@
...
 import scala.concurrent.duration.Duration
 import scala.concurrent.duration.DurationInt
 import scala.concurrent.duration.FiniteDuration
-import akka.actor.ActorSystem
-import akka.pattern.after
-import java.util.concurrent.Executors
 import scala.language.postfixOps
 
+import akka.actor.ActorSystem
+import akka.pattern.{ after => expire }
```
Comment (rabbah) [inline common/scala/src/main/scala/whisk/utils/ExecutionContextFactory.scala]: renamed `after` to `expire` to reuse `after` below for better readability on call side.

File: common/scala/src/main/scala/whisk/utils/ExecutionContextFactory.scala [hunk 3]
```diff
@@ -17,22 +17,36 @@
...
-            Future firstCompletedOf Seq(f, after(timeout, system.scheduler)(Future.failed(msg)))
+            Future firstCompletedOf Seq(f, expire(timeout, system.scheduler)(Future.failed(msg)))
+        }
+    }
+
+    /**
+     * Extends a future with an scheduled call back. The call back may be used to complete the future (with a timeout exception).
+     * The idiom to use is: promise afterTimeout(duration, promise.tryFailure(TimeoutException)`.
+     */
+    implicit class PromiseExtensions[T](f: Promise[T]) {
```
Comment (rabbah) [inline common/scala/src/main/scala/whisk/utils/ExecutionContextFactory.scala]: after `timeout` call `next`, what that does is not uninteresting for this method.
return the promise for convenience.

File: common/scala/src/main/scala/whisk/utils/ExecutionContextFactory.scala [hunk 4]
```diff
@@ -17,22 +17,36 @@
...
-            Future firstCompletedOf Seq(f, after(timeout, system.scheduler)(Future.failed(msg)))
+            Future firstCompletedOf Seq(f, expire(timeout, system.scheduler)(Future.failed(msg)))
+        }
+    }
+
+    /**
+     * Extends a future with an scheduled call back. The call back may be used to complete the future (with a timeout exception).
+     * The idiom to use is: promise afterTimeout(duration, promise.tryFailure(TimeoutException)`.
+     */
+    implicit class PromiseExtensions[T](p: Promise[T]) {
```
Comment (rabbah) [inline common/scala/src/main/scala/whisk/utils/ExecutionContextFactory.scala]: after `timeout` call `next`, what that does is not uninteresting for this method.
return the promise for convenience.

File: core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala [hunk 5]
```diff
@@ -101,9 +112,9 @@ class LoadBalancerService(config: WhiskConfig, verbosity: Verbosity.Level)(
         implicit val tid = msg.transid
         val aid = msg.response.activationId
         val response = msg.response
-        queryMap.get(aid) map { promise =>
-            promise.trySuccess(response)
-            queryMap -= aid
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala]: avoid race by removing first from the map.

File: tools/build/redo [hunk 6]
```diff
@@ -58,7 +58,7 @@ def doComponentSequence(props, args, components):
         if not component:
             if args.yaml:
                 file = c if c.endswith('.yml') else '%s.yml' % c
-                component = makeComponent('custom deployment', 'deploying using %s' % file, yaml = file)
```
Comment (rabbah) [inline tools/build/redo]: stashed but not related to this comment - general tweaks to `redo` script to support custom YAML files and added gocli build.

File: tools/cli/wskadmin [hunk 7]
```diff
@@ -86,6 +86,7 @@ def parseArgs():
 
     subcmd = subparser.add_parser('create', help='create a user and show authorization key')
     subcmd.add_argument('subject', help='the subject to create')
+    subcmd.add_argument('-u', '--auth', help='the uuid:key to initialize the subject authorization key with')
```
Comment (rabbah) [inline tools/cli/wskadmin]: stashed in this commit but not related to active ack bug - allow wskadmin to create user with a given key. validate the key is a uuid:secrete pair that conforms.

**Comment by rabbah** [comment]
Previously passed PG but since I changed implementation, new PPG 686.

File: common/scala/src/main/scala/whisk/utils/ExecutionContextFactory.scala [hunk 8]
```diff
@@ -17,22 +17,37 @@
...
         def withTimeout(timeout: FiniteDuration, msg: => Throwable)(implicit system: ActorSystem, ec: ExecutionContext): Future[T] = {
-            Future firstCompletedOf Seq(f, after(timeout, system.scheduler)(Future.failed(msg)))
+            Future firstCompletedOf Seq(f, expire(timeout, system.scheduler)(Future.failed(msg)))
+        }
+    }
+
+    /**
+     * Extends a promise with an scheduled call back. The call back may be used to complete the promise. The result of the
+     * call back is not interesting to this method.
+     * The idiom to use is: promise afterTimeout(duration, promise.tryFailure(TimeoutException)`.
```
Comment (psuter) [inline common/scala/src/main/scala/whisk/utils/ExecutionContextFactory.scala]: The comment says `afterTimeout` but it's called `after` below.

Comment (rabbah) [inline common/scala/src/main/scala/whisk/utils/ExecutionContextFactory.scala]: doh - i renamed the method and didn't update the comment; will stash in a future commit.

**Comment by rabbah** [comment]
Merged as https://github.com/openwhisk/openwhisk/commit/ee9f844a399fc3a31a132437118ceec2d39321ab.

**Pull request closed without merging**